---
title: "Software Tutorial 1 (Under Construction!)"
subtitle: "Implementation of the Slope-Hunter method on breast cancer progression"
author: "Osama Mahmoud"
date: "October 07, 2022"
site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    bibliography: sh.bib
    toc: true
    toc_depth: 2
editor_options:
  chunk_output_type: console
---

# Software Installation

## Requirements
The **SlopeHunter** software does not require any special hardware. A standard computer suitable for most applications nowadays should be sufficient.

`SlopeHunter` is written as an R package that requires `R (>= 3.5.0)` to be installed on your machine. If you do NOT have R installed, or you do have an older version of R installed on your machine, you would need to install the latest R version [from here](https://cloud.r-project.org/).

The `SlopeHunter` R package depends on a few other R packages that will be automatically installed with `SlopeHunter` when you follow the installation guide below. A list of these R dependencies is [reported here](https://github.com/Osmahmoud/SlopeHunter#r-dependencies).

## System Compatibility
The `SlopeHunter` software is compatible with *Windows*, *Linux* and *macOS*. The package has been tested in R under the following systems:

-   Linux: Ubuntu 16.04 (R 4.0.4)
-   macOS: Monterey 12.3 (R 4.2)
-   Windows: 10 (R 4.2)

## Installation Guide
To install the `SlopeHunter` R package, simply run

```{r, eval=FALSE}
# required only once per machine!
devtools::install_github("Osmahmoud/SlopeHunter")
```

# Reading Data
We are going to use data on ... (GIVE DESCRIPTION on BCa DATA)







First, you need to load the `SlopeHunter` package:

```{r}
require(SlopeHunter)
```



## Reading in the incidence GWAS

```{r}
BCa_incidence <- SlopeHunter::read_incidence(filename = "Your-incidence-data-file.gz",
  )
```















Install the *SlopeHunter* R package by following the installation instructions described [here](https://github.com/Osmahmoud/SlopeHunter).

Slope-Hunter

Report tutorials on the simulated data here ...

# Applied Data Analysis
***Example: Fasting Insulin adjusted for BMI***

## 


## Format Data for Slope-Hunter Analysis
First, you need to load the `Slope-Hunter` R package:

```{r}
library(SlopeHunter)
```

```{r, echo=FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

### Read in the incidence (the condition trait) data

```{r}
BMI_incidence <- read_incidence(filename = "Your-incidence-data-file.gz",
  gz = TRUE,
  eaf_col="Freq_Tested_Allele_in_HRS",
  effect_allele_col="Tested_Allele",
  other_allele_col="Other_Allele",
  pval_col = "P",
  pos_col = "POS")
```

### Read in the conditional outcome data (BMI-adjusted Insulin)
```{r, echo=FALSE}
Insulin_adj_BMI <- read_prognosis("ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/MahajanA_25625282_GCST007857/Mahajan_2014_SV_ExomeChip_lnFI_AdjForBMI.txt", 
                                  gz = FALSE,
                                  sep = "\t",
                                  snp_col = "ID",
                                  beta_col = "BETA",
                                  se_col = "SE",
                                  pval_col = "P",
                                  effect_allele_col = "EA",
                                  other_allele_col = "NEA",
                                  chr_col = "Chromosome",
                                  pos_col = "Position")
```


```{r}
Insulin_adj_BMI <- read_prognosis("Your-conditional-outcome-data-file.txt", 
                                  gz = FALSE,
                                  sep = "\t",
                                  snp_col = "ID",
                                  beta_col = "BETA",
                                  se_col = "SE",
                                  pval_col = "P",
                                  effect_allele_col = "EA",
                                  other_allele_col = "NEA",
                                  chr_col = "Chromosome",
                                  pos_col = "Position")
```

### Harmonisation
```{r}
Data_harmonised <- harmonise_effects(incidence_dat = BMI_incidence,
                                prognosis_dat = Insulin_adj_BMI,
                                by.pos = TRUE, pos_cols = c("POS.incidence", "POS.prognosis"),
                                snp_cols=c("SNP", "SNP"),
                                beta_cols = c("BETA.incidence", "BETA.prognosis"),
                                se_cols=c("SE.incidence", "SE.prognosis"),
                                EA_cols=c("EA.incidence", "EA.prognosis"),
                                OA_cols=c("OA.incidence", "OA.prognosis")
                                )
```

```{r}
nrow(Data_harmonised)               # No. SNPs present in both datasets
attributes(Data_harmonised)$info    # Get info on the harmonisation process
```

### Extract valid data after harmonisation (those that are eligible for pruning)
```{r}
Data_to_prune <- Data_harmonised[!Data_harmonised$remove, ]
nrow(Data_to_prune)                 # No. SNPs to be pruned
```

### Pruning data
```{r}
Data_pruned <- LD_prune(Data_to_prune, Random = TRUE, seed = 15151515)
```

## Apply the Slope-Hunter approach
TBA ...


